pkg/controller/schedulingpreference/scheduler.go

```diff
@@ -0,0 +1,148 @@
+		scheduler, err := StartSchedulingPreferenceController(kind, schedulingType.SchedulerFactory, config, fedNamespace, clusterNamespace, targetNamespace, stopChan, true)
+		if err != nil {
+			glog.Fatalf("Error starting schedulingpreference controller for %q : %v", kind, err)
+		}
+		s[kind] = scheduler
+	}
+
+	userAgent := "SchedulerController"
+	restclient.AddUserAgent(config, userAgent)
+	client := fedclientset.NewForConfigOrDie(config).CoreV1alpha1()
+
+	controller, err := newController(client, config, fedNamespace, &s)
+	if err != nil {
+		return err
+	}
+	glog.Infof("Starting scheduler controller")
+
+	controller.stopChan = stopChan
+	controller.Run(stopChan)
+
+	return nil
+}
+
+func newController(client corev1alpha1client.CoreV1alpha1Interface, config *restclient.Config, fedNamespace string, scheduler *map[string]schedulingtypes.Scheduler) (*SchedulerController, error) {
+	c := &SchedulerController{
+		scheduler: scheduler,
+	}
+
+	c.worker = util.NewReconcileWorker(c.reconcile, util.WorkerTiming{})
+
+	c.store, c.controller = cache.NewInformer(
+		&cache.ListWatch{
+			// Only watch the federation namespace to ensure
+			// restrictive authz can be applied to a namespaced
+			// control plane.
+			ListFunc: func(options metav1.ListOptions) (pkgruntime.Object, error) {
+				return client.FederatedTypeConfigs(fedNamespace).List(options)
+			},
+			WatchFunc: func(options metav1.ListOptions) (watch.Interface, error) {
+				return client.FederatedTypeConfigs(fedNamespace).Watch(options)
+			},
+		},
+		&corev1a1.FederatedTypeConfig{},
+		util.NoResyncPeriod,
+		util.NewTriggerOnAllChanges(c.worker.EnqueueObject),
+	)
+
+	return c, nil
+}
+
+// Run runs the Controller.
+func (c *SchedulerController) Run(stopChan <-chan struct{}) {
+	go c.controller.Run(stopChan)
+
+	// wait for the caches to synchronize before starting the worker
+	if !cache.WaitForCacheSync(stopChan, c.controller.HasSynced) {
+		runtime.HandleError(fmt.Errorf("Timed out waiting for cache to sync"))
+		return
+	}
+
+	c.worker.Run(stopChan)
+}
+
+func (c *SchedulerController) reconcile(qualifiedName util.QualifiedName) util.ReconciliationStatus {
+	key := qualifiedName.String()
+
+	glog.Infof("Running reconcile FederatedTypeConfig for %q", key)
+
+	cachedObj, exist, err := c.store.GetByKey(key)
+	if err != nil {
+		runtime.HandleError(fmt.Errorf("Failed to query FederatedTypeConfig store for %q: %v", key, err))
+		return util.StatusError
+	}
+	if !exist {
+		return util.StatusAllOK
+	}
+	typeConfig := cachedObj.(*corev1a1.FederatedTypeConfig)
+
+	deleted := typeConfig.DeletionTimestamp != nil
+	if deleted {
+		// Do not support deletion yet
+		return util.StatusAllOK
+	}
+
+	// set name and group for the type config target
+	corev1a1.SetFederatedTypeConfigDefaults(typeConfig)
+	apiResource := typeConfig.GetTarget()
+	kind := typeConfig.GetTemplate().Kind
+
+	for name, _ := range schedulingtypes.SchedulingTypes() {
```

Comment:
I don't think it should be valid to attempt to register plugins for all type configs.  Consider adding a mapping of kind to scheduling type and only register kind plugins with an appropriate scheduler.  This will also aid in discovering which scheduler to stop a plugin for when the `FederatedTypeConfig` goes away or has propagation disabled.

@irfanurrehman Should scheduling be undertaken for a type even if propagation is disabled?  Is a separate 'schedulingEnabled' option suggested?

Reply:
I don't think there is a need of another option. There already is a feature flag `features.SchedulerPreferences`, enabling which enables the high level scheduling. If there is a need in future to allow users to enable/disable schedulers at more fine grained level, we can break this into sub features. 
It is quite appropriate to start/stop a given high level scheduler based on the propagation enabled/disabled for the target type. As you suggested a target kind to a scheduler map would be useful. This also indicates that an overall `Stop` routine for the scheduler wrt a given type is necessary right now (which seems to be missing).

We can consider another additional point (may be in some future PR, might be an overkill for now) is to check for the target type CRD being installed/not installed and then start controllers in response.